src/scheduler/pbs_sched_utils.cpp

```diff
@@ -684,13 +744,56 @@ sched_svr_init(void)
  * @return void
  */
 static void
-reconnect_servers()
+reconnect_servers(void)
 {
+	svr_conn_t **prim_conn = get_conn_svr_instances(clust_primary_sock);
+	svr_conn_t **sec_conn = get_conn_svr_instances(clust_secondary_sock);
+	int num_down = 0;
+	static time_t last_reconnect = -1;
+	const int min_reconnect_period = 30;	/* Reconnect at least 30secs after last */
+
 	pthread_mutex_lock(&cleanup_lock);
+	if (prim_conn == NULL || sec_conn == NULL || get_num_servers() == 1) {
+		close_servers();
+		connect_svrpool();
+		pthread_mutex_unlock(&cleanup_lock);
 
-	close_servers();
-	connect_svrpool();
+		return;
+	}
+
+	part_tolerance = true;
+
+	if (time(NULL) - last_reconnect < min_reconnect_period) {
+		return;	/* To throttle the reconnect frequency, don't connect if under min_reconnect_period */
```

Comment:
Unlock the mutex. We should consider making a "Locker" class that takes a given mutex in the constructor and releases it in the destructor. That way we don't have to worry about forgetting to release mutexes.

Reply:
Great idea, but as we discussed, let's wait until we refactor code to use std:thread and std:mutex, then we can use STL's std::lock_guard or std::unique_lock to do this.